Cache results from Imagick::queryFormats

joemcgill opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/61532

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

joemcgill committed:
Cache results from Imagick::queryFormats

joemcgill committed:
Create initial supports array.

joemcgill committed:
Space invaders

adamsilverstein commented:
Great idea, love it! Curious how you came to the realization that the mime type check call was slow/expensive? Did you see it in a trace somewhere?

westonruter commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -107,15 +107,30 @@ public static function supports_mime_type( $mime_type ) {
+		// Imagick::queryFormats is not performant, so cache the results.
+		$supports = wp_cache_get( 'imagick_supports' );
+
+		if ( ! $supports ) {
```
Since technically any type could be stored in the cache value:

```suggestion
		if ( ! is_array( $supports ) ) {
```

felixarntz commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -107,15 +107,30 @@ public static function supports_mime_type( $mime_type ) {
+		}
+
+		// Imagick::queryFormats is not performant, so cache the results.
+		$supports = wp_cache_get( 'imagick_supports' );
```
We shouldn't use the object cache without a group. Not sure what would be most fitting for this use-case, but I think we should try to find one.

westonruter commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -107,15 +107,30 @@ public static function supports_mime_type( $mime_type ) {
+			$supports = array();
+		}
+
+		if ( isset( $supports[ $imagick_extension ] ) ) {
```
Hardening, although this function doesn't currently have a strict type hint return value:

```suggestion
		if ( isset( $supports[ $imagick_extension ] ) && is_bool( $supports[ $imagick_extension ] ) ) {
```

westonruter commented on src/wp-includes/class-wp-image-editor-imagick.php:
```diff
@@ -107,15 +107,30 @@ public static function supports_mime_type( $mime_type ) {
+		}
+
+		if ( isset( $supports[ $imagick_extension ] ) ) {
+			return $supports[ $imagick_extension ];
```
Alternatively:

```suggestion
			return (bool) $supports[ $imagick_extension ];
```

felixarntz reviewed (commented):
@joemcgill This is a good idea IMO, though at first glance I think this cache would apply at a _very_ low level.

I think a more comprehensive solution would be to cache part of what `_wp_image_editor_choose()` does, for example:
* After the `wp_image_editors` filter, hash the array of classes and `$args` and use that in the cache key.
* This cache would then avoid the entire `foreach` loop, including any other potentially expensive check methods from other image editor class implementations.
* The values of the filter probably barely ever changes, but by using a hash we can ensure no stale value would be served.

WDYT?

adamsilverstein commented:
> I think a more comprehensive solution would be to cache part of what _wp_image_editor_choose() does, for example:

Interesting idea to add caching up at this level, although this PR's changes could still be useful for `supports_mime_type` since that could be called from elsewhere including plugins. 

My one concern would be ensuring the cache has some expiration so we can catch new capabilities when users upgrade their systems. This brings me back to wondering how significant this improvement is - does profiling indicate this function is problematic?

felixarntz commented:
> My one concern would be ensuring the cache has some expiration so we can catch new capabilities when users upgrade their systems. This brings me back to wondering how significant this improvement is - does profiling indicate this function is problematic?

This is a good point, would be good to clarify _how_ bad the performance of the uncached function makes it. And related to expiration, I wonder: @joemcgill Are you thinking this should be persistently cached, or would it already benefit to be cached non-persistently (e.g. because the function is called many times during a single page load)?

mukeshpanchal27 commented:
Why aren't the unit tests and other GitHub jobs working for this PR?

desrosj committed:
Merge remote-tracking branch 'upstream/trunk' into fix/61532-cache-imagick-mime-supports

desrosj commented:
> Why aren't the unit tests and other GitHub jobs working for this PR?

I've seen this happen before if the PR is opened with a base branch that does not contain GitHub Action workflows (such as the old `master` branch).

It does not show that the base branch for the PR was changed after opening, but I've gone and merged the current state of `trunk` into this branch and it seems like everything is now running!

joemcgill committed:
Revert changes to Imagick editor

joemcgill committed:
Cache wp_image_editor_supports results

joemcgill committed:
Add new image_editor global cache group

joemcgill commented:
@adamsilverstein:

> Great idea, love it! Curious how you came to the realization that the mime type check call was slow/expensive? Did you see it in a trace somewhere?

Was alerted to the issue while checkout out the XHProf flamegraphs demo from @joehoyle here:https://x.com/joe_hoyle/status/1806347618185875819

@felixarntz:
> This is a good idea IMO, though at first glance I think this cache would apply at a very low level.

That's a good observation. I was trying to specifically target this one known issue that was impacting the editor. However, I think caching at a higher level makes sense. I've updated this PR to cache the results of the `wp_image_editor_supports` check, which is the function that is that causes `_wp_image_editor_choose` to be called most often (4 times during a normal editor load).

<img width="1502" alt="image" src="https://github.com/user-attachments/assets/52ae37fa-f65e-4c7f-bb4d-9592c725ea27">

With this PR applied, `_wp_image_editor_choose` never gets called with a warm cache.

I've added a new global cache group, called 'image_editor', and store this value to a `wp_image_editor_supports` cache key, with a TTL of 1 day.

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -4096,7 +4096,21 @@ function wp_get_image_editor( $path, $args = array() ) {
+		$supports = array();
+	}
+
+	$cache_key = md5( serialize( $args ) );
```
I think the note from my previous comment https://github.com/WordPress/wordpress-develop/pull/6936#pullrequestreview-2221590113 is relevant here: Caching `$args` alone is not enough to ensure this cache doesn't serve stale values. We need to also account for the `wp_image_editors` filter output, otherwise someone that installs a plugin that uses this filter won't see the effects. For example, that filter is used in the "Image Placeholders" plugin we have.

I would also argue that caching within `_wp_image_editor_choose()` makes more sense as that function is called in other places too, and the way that it works doesn't change for these other callers. So by caching the result of `_wp_image_editor_choose()` we expand the reach of this enhancement.

In any case, the main problem here is not accounting for the filter in the cache key.

joemcgill replied:
[6056c51](https://github.com/WordPress/wordpress-develop/pull/6936/commits/6056c51f2879bbdd63f0b289de90d1b356d0294e)

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -876,6 +876,7 @@ function wp_start_object_cache() {
 				'blog-lookup',
 				'blog_meta',
 				'global-posts',
+				'image_editor',
```
+1 to using a global cache group.

felixarntz reviewed (commented):
@joemcgill I think the cache is missing one crucial detail, but otherwise it's looking close.

joemcgill committed:
Revert "Cache wp_image_editor_supports results"

This reverts commit a86a6181953ef5dfb665ac67ddf9af8818b5e120.

joemcgill committed:
Cache result of _wp_image_editor_choose

joemcgill commented:
Thanks @felixarntz. [Good point about the implementation filter](https://github.com/WordPress/wordpress-develop/pull/6936#discussion_r1787983127). Also, I was originally thinking just caching the supports check would be safer, but caching the implementation does have broader potential impact. I've updated this PR to move the cache to `_wp_image_editor_choose()` and the results are looking pretty good.

Here's my local profile of the `/wp-admin/post-new.php` page before and after the cache is warmed. Almost all of the savings is still coming from avoiding calls to  `WP_Image_Editor_Imagick::supports_mime_type` but this is a bit broader and should have a noticeable impact on editor loading times on systems with a persistent object cache and may also speed up image uploads.

## Cold cache
<img width="1496" alt="image" src="https://github.com/user-attachments/assets/061d4a78-1fe8-4584-b14d-ee043af05459">

## Warm cache
<img width="1506" alt="image" src="https://github.com/user-attachments/assets/695bc238-7db4-4905-8c64-2f31a553f03f">

felixarntz commented on src/wp-includes/media.php:
```diff
@@ -4123,7 +4123,22 @@ function _wp_image_editor_choose( $args = array() ) {
 	$implementations = apply_filters( 'wp_image_editors', array( 'WP_Image_Editor_Imagick', 'WP_Image_Editor_GD' ) );
-	$supports_input  = false;
+
+	$editors = wp_cache_get( 'wp_image_editor_choose', 'image_editor' );
+
+	if ( ! is_array( $editors ) ) {
+		$editors = array();
+	}
+
+	// Cache the chosen editor implementation based on specific args and available implementations.
+	$cache_key = md5( serialize( array( $args, $implementations ) ) );
+
+	if ( isset( $editors[ $cache_key ] ) ) {
```
I'm curious what's the rationale for nesting all the cache values in a single cache key "bucket". In most places in Core we store individual values per cache key, and the dynamic portion becomes part of the actual cache key.

For example the actual cache key would be:
```php
$cache_key = 'wp_image_editor_choose_' . md5( serialize( array( $args, $implementations ) ) );
```

Since these caches would expire after 1 day anyway, I don't think there's any harm in storing in individual cache keys. This approach here might work too, but there's potential concerns about scalability (if too many different combinations of arguments are all stored under a single cache key) and diverging from how I believe we typically handle such caches in Core.

joemcgill replied:
I didn't have any real intention when I chose to implement it this way, but I can see a few benefits. 

Generally, including the hash as part of the name is done for the purpose of avoiding stale caches (i.e., if the values that are hashed change, the `wp_get_cache()` call will result in a cache miss). That's not really a concern for this cache because the need to store the data for long periods while avoiding returning stale values is not that high.

This specific function can get called multiple times during the same run with different args. Having the results stored in multiple distinct caches would mean that we need to load separate cache values for each specific set of args. By saving all of the values to one cache, all of the cached values are loaded to memory the first time `_wp_image_editor_choose()` is called, and can be accessed from memory on each subsequent call.

Another minor benefit is that it makes it a bit easier to work with a cache that has a static name in the DB instead of a dynamic one. As a theoretical example you could do something like `wp cache get wp_image_editor_choose image_editor` or `wp cache delete wp_image_editor_choose image_editor` to review or remove the values, which is more difficult with a dynamic name.

felixarntz replied:
That sounds good to me, makes sense. Let's go with this approach then.

felixarntz reviewed (commented):
@joemcgill This implementation looks good to me and would work. I have one question on the approach that I missed to spot before.

felixarntz reviewed (approved):
LGTM!

joemcgill commented:
Committed in https://core.trac.wordpress.org/changeset/59189.
